src/jltypes.c

```diff
@@ -1015,15 +1015,15 @@ struct _jl_typestack_t;
 typedef struct _jl_typestack_t jl_typestack_t;
 
 static jl_value_t *inst_datatype_inner(jl_datatype_t *dt, jl_svec_t *p, jl_value_t **iparams, size_t ntp,
-                                       jl_typestack_t *stack, jl_typeenv_t *env);
+                                       jl_typestack_t *stack, jl_typeenv_t *env, int check);
 
 // Build an environment mapping a TypeName's parameters to parameter values.
 // This is the environment needed for instantiating a type's supertype and field types.
 static jl_value_t *inst_datatype_env(jl_value_t *dt, jl_svec_t *p, jl_value_t **iparams, size_t ntp,
                                      jl_typestack_t *stack, jl_typeenv_t *env, int c)
 {
     if (jl_is_datatype(dt))
-        return inst_datatype_inner((jl_datatype_t*)dt, p, iparams, ntp, stack, env);
+        return inst_datatype_inner((jl_datatype_t*)dt, p, iparams, ntp, stack, env, 1);
```

Comment:
I think we could potentially relax this further (currently) to `all(isa(TypeVar), iparams)`, since we don't check TypeVar? We might need to make sure that ub and lb is unchanged also (so that a union won't become degenerate for example)

Reply:
And `iparams` must be all unique? Do we have this pattern in Compiler? If not, I guess the gain won't be large.

Reply:
Also true. It probably is rare, so likely doesn't matter as you say